{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f8d57862-99f5-4ff2-84eb-7746db4e9edc",
   "metadata": {},
   "source": [
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception.  \n",
    "Exceptions are just regular classes that inherit from the Exception class. All exception classes are derived from the BaseException class.\n",
    "The code can run built in exceptions, or we can also raise these exceptions in the code. User can derive their own exception from the Exception class, or from any other child class of Exception class.\n",
    "The BaseException is the base class of all other exceptions. User defined classes cannot be directly derived from this class, to derive user defied class, we need to use Exception class.\n",
    "To create a custom exception class, we define a class that inherits from the built-in Exception class"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277b023a-3aa6-4de6-b3c8-0e916ebb3bb3",
   "metadata": {},
   "source": [
    "Q2. Write a python program to print Python Exception Hierarchy.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7d864f0b-2daf-4595-a4c0-5ce537bfe677",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Hierarchy for python in-built exceptions: \n",
      " BaseException\n",
      "--- Exception\n",
      "------ TypeError\n",
      "--------- FloatOperation\n",
      "--------- MultipartConversionError\n",
      "------ StopAsyncIteration\n",
      "------ StopIteration\n",
      "------ ImportError\n",
      "--------- ModuleNotFoundError\n",
      "--------- ZipImportError\n",
      "------ OSError\n",
      "--------- ConnectionError\n",
      "------------ BrokenPipeError\n",
      "------------ ConnectionAbortedError\n",
      "------------ ConnectionRefusedError\n",
      "------------ ConnectionResetError\n",
      "--------------- RemoteDisconnected\n",
      "--------- BlockingIOError\n",
      "--------- ChildProcessError\n",
      "--------- FileExistsError\n",
      "--------- FileNotFoundError\n",
      "--------- IsADirectoryError\n",
      "--------- NotADirectoryError\n",
      "--------- InterruptedError\n",
      "------------ InterruptedSystemCall\n",
      "--------- PermissionError\n",
      "--------- ProcessLookupError\n",
      "--------- TimeoutError\n",
      "--------- UnsupportedOperation\n",
      "--------- itimer_error\n",
      "--------- herror\n",
      "--------- gaierror\n",
      "--------- SSLError\n",
      "------------ SSLCertVerificationError\n",
      "------------ SSLZeroReturnError\n",
      "------------ SSLWantWriteError\n",
      "------------ SSLWantReadError\n",
      "------------ SSLSyscallError\n",
      "------------ SSLEOFError\n",
      "--------- Error\n",
      "------------ SameFileError\n",
      "--------- SpecialFileError\n",
      "--------- ExecError\n",
      "--------- ReadError\n",
      "--------- URLError\n",
      "------------ HTTPError\n",
      "------------ ContentTooShortError\n",
      "--------- BadGzipFile\n",
      "------ EOFError\n",
      "--------- IncompleteReadError\n",
      "------ RuntimeError\n",
      "--------- RecursionError\n",
      "--------- NotImplementedError\n",
      "------------ ZMQVersionError\n",
      "------------ StdinNotImplementedError\n",
      "--------- _DeadlockError\n",
      "--------- BrokenBarrierError\n",
      "--------- BrokenExecutor\n",
      "------------ BrokenThreadPool\n",
      "--------- SendfileNotAvailableError\n",
      "--------- ExtractionError\n",
      "--------- VariableError\n",
      "------ NameError\n",
      "--------- UnboundLocalError\n",
      "------ AttributeError\n",
      "--------- FrozenInstanceError\n",
      "------ SyntaxError\n",
      "--------- IndentationError\n",
      "------------ TabError\n",
      "------ LookupError\n",
      "--------- IndexError\n",
      "--------- KeyError\n",
      "------------ NoSuchKernel\n",
      "------------ UnknownBackend\n",
      "--------- CodecRegistryError\n",
      "------ ValueError\n",
      "--------- UnicodeError\n",
      "------------ UnicodeEncodeError\n",
      "------------ UnicodeDecodeError\n",
      "------------ UnicodeTranslateError\n",
      "--------- UnsupportedOperation\n",
      "--------- JSONDecodeError\n",
      "--------- SSLCertVerificationError\n",
      "--------- Error\n",
      "--------- UnsupportedDigestmodError\n",
      "--------- IllegalMonthError\n",
      "--------- IllegalWeekdayError\n",
      "--------- ParserError\n",
      "--------- ClassNotFound\n",
      "--------- ClipboardEmpty\n",
      "--------- MessageDefect\n",
      "------------ NoBoundaryInMultipartDefect\n",
      "------------ StartBoundaryNotFoundDefect\n",
      "------------ CloseBoundaryNotFoundDefect\n",
      "------------ FirstHeaderLineIsContinuationDefect\n",
      "------------ MisplacedEnvelopeHeaderDefect\n",
      "------------ MissingHeaderBodySeparatorDefect\n",
      "------------ MultipartInvariantViolationDefect\n",
      "------------ InvalidMultipartContentTransferEncodingDefect\n",
      "------------ UndecodableBytesDefect\n",
      "------------ InvalidBase64PaddingDefect\n",
      "------------ InvalidBase64CharactersDefect\n",
      "------------ InvalidBase64LengthDefect\n",
      "------------ HeaderDefect\n",
      "--------------- InvalidHeaderDefect\n",
      "--------------- HeaderMissingRequiredValue\n",
      "--------------- NonPrintableDefect\n",
      "--------------- ObsoleteHeaderDefect\n",
      "--------------- NonASCIILocalPartDefect\n",
      "--------------- InvalidDateDefect\n",
      "--------- MacroToEdit\n",
      "--------- InvalidFileException\n",
      "--------- UnequalIterablesError\n",
      "--------- InvalidVersion\n",
      "--------- _InvalidELFFileHeader\n",
      "--------- InvalidWheelFilename\n",
      "--------- InvalidSdistFilename\n",
      "--------- InvalidSpecifier\n",
      "--------- InvalidMarker\n",
      "--------- UndefinedComparison\n",
      "--------- UndefinedEnvironmentName\n",
      "--------- InvalidRequirement\n",
      "------------ RequirementParseError\n",
      "--------- InvalidVersion\n",
      "------ AssertionError\n",
      "------ ArithmeticError\n",
      "--------- FloatingPointError\n",
      "--------- OverflowError\n",
      "--------- ZeroDivisionError\n",
      "------------ DivisionByZero\n",
      "------------ DivisionUndefined\n",
      "--------- DecimalException\n",
      "------------ Clamped\n",
      "------------ Rounded\n",
      "--------------- Underflow\n",
      "--------------- Overflow\n",
      "------------ Inexact\n",
      "--------------- Underflow\n",
      "--------------- Overflow\n",
      "------------ Subnormal\n",
      "--------------- Underflow\n",
      "------------ DivisionByZero\n",
      "------------ FloatOperation\n",
      "------------ InvalidOperation\n",
      "--------------- ConversionSyntax\n",
      "--------------- DivisionImpossible\n",
      "--------------- DivisionUndefined\n",
      "--------------- InvalidContext\n",
      "------ SystemError\n",
      "--------- CodecRegistryError\n",
      "------ ReferenceError\n",
      "------ MemoryError\n",
      "------ BufferError\n",
      "------ Warning\n",
      "--------- UserWarning\n",
      "------------ GetPassWarning\n",
      "------------ FormatterWarning\n",
      "--------- EncodingWarning\n",
      "--------- DeprecationWarning\n",
      "------------ ProvisionalWarning\n",
      "--------- PendingDeprecationWarning\n",
      "--------- SyntaxWarning\n",
      "--------- RuntimeWarning\n",
      "------------ ProactorSelectorThreadWarning\n",
      "------------ UnknownTimezoneWarning\n",
      "------------ PEP440Warning\n",
      "--------- FutureWarning\n",
      "------------ ProvisionalCompleterWarning\n",
      "--------- ImportWarning\n",
      "--------- UnicodeWarning\n",
      "--------- BytesWarning\n",
      "--------- ResourceWarning\n",
      "--------- DeprecatedTzFormatWarning\n",
      "--------- PkgResourcesDeprecationWarning\n",
      "------ _OptionError\n",
      "------ _Error\n",
      "------ error\n",
      "------ Verbose\n",
      "------ Error\n",
      "------ SubprocessError\n",
      "--------- CalledProcessError\n",
      "--------- TimeoutExpired\n",
      "------ TokenError\n",
      "------ StopTokenizing\n",
      "------ ClassFoundException\n",
      "------ EndOfBlock\n",
      "------ TraitError\n",
      "------ Error\n",
      "------ Error\n",
      "--------- CancelledError\n",
      "--------- TimeoutError\n",
      "--------- InvalidStateError\n",
      "------ _GiveupOnSendfile\n",
      "------ error\n",
      "------ Incomplete\n",
      "------ TimeoutError\n",
      "------ InvalidStateError\n",
      "------ LimitOverrunError\n",
      "------ QueueEmpty\n",
      "------ QueueFull\n",
      "------ Empty\n",
      "------ Full\n",
      "------ ArgumentError\n",
      "------ ZMQBaseError\n",
      "--------- ZMQError\n",
      "------------ ContextTerminated\n",
      "------------ Again\n",
      "------------ InterruptedSystemCall\n",
      "--------- ZMQBindError\n",
      "--------- NotDone\n",
      "------ PickleError\n",
      "--------- PicklingError\n",
      "--------- UnpicklingError\n",
      "------ _Stop\n",
      "------ ArgumentError\n",
      "------ ArgumentTypeError\n",
      "------ ConfigError\n",
      "--------- ConfigLoaderError\n",
      "------------ ArgumentError\n",
      "--------- ConfigFileNotFound\n",
      "------ ConfigurableError\n",
      "--------- MultipleInstanceError\n",
      "------ ApplicationError\n",
      "------ error\n",
      "------ TimeoutError\n",
      "------ error\n",
      "------ ReturnValueIgnoredError\n",
      "------ KeyReuseError\n",
      "------ UnknownKeyError\n",
      "------ LeakedCallbackError\n",
      "------ BadYieldError\n",
      "------ ReturnValueIgnoredError\n",
      "------ Return\n",
      "------ InvalidPortNumber\n",
      "------ error\n",
      "------ LZMAError\n",
      "------ RegistryError\n",
      "------ _GiveupOnFastCopy\n",
      "------ Error\n",
      "--------- NoSectionError\n",
      "--------- DuplicateSectionError\n",
      "--------- DuplicateOptionError\n",
      "--------- NoOptionError\n",
      "--------- InterpolationError\n",
      "------------ InterpolationMissingOptionError\n",
      "------------ InterpolationSyntaxError\n",
      "------------ InterpolationDepthError\n",
      "--------- ParsingError\n",
      "------------ MissingSectionHeaderError\n",
      "------ NoIPAddresses\n",
      "------ BadZipFile\n",
      "------ LargeZipFile\n",
      "------ BadEntryPoint\n",
      "------ NoSuchEntryPoint\n",
      "------ DuplicateKernelError\n",
      "------ ErrorDuringImport\n",
      "------ NotOneValueFound\n",
      "------ CannotEval\n",
      "------ OptionError\n",
      "------ BdbQuit\n",
      "------ Restart\n",
      "------ ExceptionPexpect\n",
      "--------- EOF\n",
      "--------- TIMEOUT\n",
      "------ PtyProcessError\n",
      "------ FindCmdError\n",
      "------ HomeDirError\n",
      "------ ProfileDirError\n",
      "------ IPythonCoreError\n",
      "--------- TryNext\n",
      "--------- UsageError\n",
      "--------- StdinNotImplementedError\n",
      "------ InputRejected\n",
      "------ GetoptError\n",
      "------ ErrorToken\n",
      "------ PrefilterError\n",
      "------ AliasError\n",
      "--------- InvalidAliasError\n",
      "------ Error\n",
      "--------- InterfaceError\n",
      "--------- DatabaseError\n",
      "------------ InternalError\n",
      "------------ OperationalError\n",
      "------------ ProgrammingError\n",
      "------------ IntegrityError\n",
      "------------ DataError\n",
      "------------ NotSupportedError\n",
      "------ Warning\n",
      "------ SpaceInInput\n",
      "------ DOMException\n",
      "--------- IndexSizeErr\n",
      "--------- DomstringSizeErr\n",
      "--------- HierarchyRequestErr\n",
      "--------- WrongDocumentErr\n",
      "--------- InvalidCharacterErr\n",
      "--------- NoDataAllowedErr\n",
      "--------- NoModificationAllowedErr\n",
      "--------- NotFoundErr\n",
      "--------- NotSupportedErr\n",
      "--------- InuseAttributeErr\n",
      "--------- InvalidStateErr\n",
      "--------- SyntaxErr\n",
      "--------- InvalidModificationErr\n",
      "--------- NamespaceErr\n",
      "--------- InvalidAccessErr\n",
      "--------- ValidationErr\n",
      "------ ValidationError\n",
      "------ EditReadOnlyBuffer\n",
      "------ _Retry\n",
      "------ InvalidLayoutError\n",
      "------ HeightIsUnknownError\n",
      "------ ParserSyntaxError\n",
      "------ InternalParseError\n",
      "------ _PositionUpdatingFinished\n",
      "------ SimpleGetItemNotFound\n",
      "------ UncaughtAttributeError\n",
      "------ HasNoContext\n",
      "------ ParamIssue\n",
      "------ _JediError\n",
      "--------- InternalError\n",
      "--------- WrongVersion\n",
      "--------- RefactoringError\n",
      "------ OnErrorLeaf\n",
      "------ InvalidPythonEnvironment\n",
      "------ MessageError\n",
      "--------- MessageParseError\n",
      "------------ HeaderParseError\n",
      "------------ BoundaryError\n",
      "--------- MultipartConversionError\n",
      "--------- CharsetError\n",
      "------ Error\n",
      "------ HTTPException\n",
      "--------- NotConnected\n",
      "--------- InvalidURL\n",
      "--------- UnknownProtocol\n",
      "--------- UnknownTransferEncoding\n",
      "--------- UnimplementedFileMode\n",
      "--------- IncompleteRead\n",
      "--------- ImproperConnectionState\n",
      "------------ CannotSendRequest\n",
      "------------ CannotSendHeader\n",
      "------------ ResponseNotReady\n",
      "--------- BadStatusLine\n",
      "------------ RemoteDisconnected\n",
      "--------- LineTooLong\n",
      "------ InteractivelyDefined\n",
      "------ KillEmbedded\n",
      "------ Error\n",
      "--------- NoSuchProcess\n",
      "------------ ZombieProcess\n",
      "--------- AccessDenied\n",
      "--------- TimeoutExpired\n",
      "------ _Ipv6UnsupportedError\n",
      "------ QueueEmpty\n",
      "------ QueueFull\n",
      "------ DebuggerInitializationError\n",
      "------ ExpatError\n",
      "------ Error\n",
      "--------- ProtocolError\n",
      "--------- ResponseError\n",
      "--------- Fault\n",
      "------ ParseBaseException\n",
      "--------- ParseException\n",
      "--------- ParseFatalException\n",
      "------------ ParseSyntaxException\n",
      "------ RecursiveGrammarException\n",
      "------ ResolutionError\n",
      "--------- VersionConflict\n",
      "------------ ContextualVersionConflict\n",
      "--------- DistributionNotFound\n",
      "--------- UnknownExtra\n",
      "------ _Error\n",
      "------ UnableToResolveVariableException\n",
      "------ InvalidTypeInArgsException\n",
      "--- GeneratorExit\n",
      "--- SystemExit\n",
      "--- KeyboardInterrupt\n",
      "--- CancelledError\n",
      "--- AbortThread\n"
     ]
    }
   ],
   "source": [
    "import inspect as ipt\n",
    "def tree_class(cls, ind=0):\n",
    "    print(\"-\"* ind,cls.__name__)\n",
    "    for K in cls.__subclasses__():\n",
    "        tree_class(K,ind+3)\n",
    "print (\"The Hierarchy for python in-built exceptions: \")\n",
    "ipt.getclasstree(ipt.getmro(BaseException)) \n",
    "tree_class(BaseException)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a890c65-3893-40a5-80e6-6488d57a5134",
   "metadata": {},
   "source": [
    "Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.  \n",
    "ArithmeticError is simply an error that occurs during numeric calculations.  \n",
    "\n",
    "ArithmeticError types in Python include:  \n",
    "\n",
    "OverFlowError  \n",
    "ZeroDivisionError  \n",
    "FloatingPointError  \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67477e3a-aea3-4167-b6ca-d9bcba5f7abb",
   "metadata": {},
   "source": [
    "##OverFlowError  \n",
    "Python OverflowError is raised when the result of a numeric calculation is too large to store.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9ae0e820-81d7-4914-baed-8fdb6954ce09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(34, 'Numerical result out of range')\n"
     ]
    }
   ],
   "source": [
    "n = 8.0\n",
    "try: \n",
    "    for i in range(1, 100):\n",
    "        n = n ** i\n",
    "except OverflowError as e:\n",
    "    print(\"OverflowError\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "186c5952-bfc8-4ab5-9dc3-b44cf34fa728",
   "metadata": {},
   "source": [
    "##ZeroDivisionError\n",
    "-Division by zero  \n",
    "Mathematically, dividing an integer by zero is wrong, and that is the reason why Python crashes the program and returns an error message."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ffc84e4d-8bb3-4ff9-bd08-1541772f86f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    arithmetic = 5/0\n",
    "    print(arithmetic)\n",
    "except ZeroDivisionError as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f2b566-53dd-4594-8119-40af4fefec2d",
   "metadata": {},
   "source": [
    "Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.  \n",
    "The LookupError exception in Python forms the base class for all exceptions that are raised when an index or a key is not found for a sequence or dictionary respectively.  \n",
    "We can use LookupError exception class to handle both IndexError and KeyError exception classes.  \n",
    "##Keyerror  \n",
    "A Python KeyError exception is what is raised when you try to access a key that isn’t in a dictionary (dict)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee0b8c45-cb43-4e12-b0d5-5568eec6f70d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Key not found 'Papaya'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    Fruits={\"apple\": 100, \"banana\": 20, \"Mango\": 75}\n",
    "    print(Fruits[\"Papaya\"])\n",
    "except KeyError as e:\n",
    "    print(\"Key not found\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ae1e44e-7f28-4034-96ba-f10de88e31f9",
   "metadata": {},
   "source": [
    "##Index error  \n",
    "\n",
    "IndexError is an exception in python that occurs when we try to access an element from a list or tuple from an index that is not present in the list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "65b3fa27-0934-4550-9fa3-ce3caae6ae67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Its an Indexerror:  list index out of range\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    a=[1,2,3,4,5,6]\n",
    "    print(a[6])\n",
    "except IndexError as e:\n",
    "    print(\"Its an Indexerror: \", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a465917-a080-4b62-b0d4-3fc0a114f481",
   "metadata": {},
   "source": [
    "Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "845ab546-f2fd-4d8e-981f-76cd6c58e18d",
   "metadata": {},
   "source": [
    "##ImportError:  \n",
    "This error generally occurs when a class cannot be imported due to one of the following reasons:  \n",
    "             -The imported class is in a circular dependency.  \n",
    "             -The imported class is unavailable or was not created.  \n",
    "             -The imported class name is misspelled.  \n",
    "             -The imported class from a module is misplaced.  \n",
    "             -The imported class is unavailable in the Python library.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5e54322e-0e6d-4451-8ea4-169d81a841c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Import error:  cannot import name 'reduced' from 'functools' (/opt/conda/lib/python3.10/functools.py)\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    a=[1,2,3,4,5,6]\n",
    "    from functools import reduced\n",
    "    result= reduced(lambda x,y :x*y, a)\n",
    "    print(result)\n",
    "except ImportError as e:\n",
    "    print(\"Import error: \", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a30b3001-c6e4-4ef8-a761-2655bc4a7d96",
   "metadata": {},
   "source": [
    "##ModuleNotFoundError:  \n",
    "The ModuleNotFoundError: No module named error is raised when Python either cannot find the module you're trying to import, the name of the package being imported was misspelled,  \n",
    "or the module doesn't exist on your hard drive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bf4b42cd-d3f4-4471-92f1-575d51af58da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ModuleNotFoundError :  No module named 'funtools'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    a=[1,2,3,4,5,6]\n",
    "    from funtools import reduce\n",
    "    result= reduce(lambda x,y :x*y, a)\n",
    "    print(result)\n",
    "except ModuleNotFoundError as e:\n",
    "    print(\"ModuleNotFoundError : \", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a2bf17c-0f85-4df8-a991-018cfc1a87a5",
   "metadata": {},
   "source": [
    "Q6. List down some best practices for exception handling in python.   \n",
    "\n",
    "1. Catch Specific Exceptions:  \n",
    "    When we catch a general exception, like Exception, we’re catching everything. This includes system-level errors that are unlikely to be handled gracefully by our code. It’s better to be explicit about which exceptions we want to catch, so we can handle them appropriately.\n",
    "\n",
    "2. Always Clean Up Resources in a Finally Block:  \n",
    "    Suppose we have a file that we need to open, read from, and then close. If an exception occurs while reading from the file, we’ll want to make sure the file is properly closed before moving on. Otherwise, we risk leaving the file in an inconsistent state or even corrupting it.\n",
    "    The best way to ensure the file is properly closed is to put the code for closing it in a finally block. That way, whether or not an exception occurs, the file will always be closed before the program continues.\n",
    "\n",
    "3. Avoid Raising Generic Exceptions:  \n",
    "    When we raise a generic exception, such as Exception or RuntimeError, we are essentially saying “I don’t know what went wrong, but something did.” This is not helpful for either us or our users. It’s much better to be specific about the error that occurred.\n",
    "\n",
    "4. Raise Custom Exceptions: \n",
    "    When we’re writing code, it’s important to think about what could go wrong and plan for those contingencies. That way, if something does go wrong, our code can gracefully handle the error instead of crashing.\n",
    "    One way to do this is to raise custom exceptions. By raising a custom exception, we can provide a specific error message that will be displayed to the user. This is helpful because it can give the user a clue as to what went wrong and how to fix it.\n",
    "\n",
    "5. Document All Exceptions Thrown by a Function:  \n",
    "    If we don’t document the exceptions thrown by a function, then other developers who use that function won’t know what to expect. This can lead to unexpected behavior, and can even cause errors in production if an exception is raised that wasn’t anticipated.  \n",
    "     Documenting exceptions is also important for code maintainability. If we need to change the way a function behaves when an exception is raised, we’ll need to update the documentation accordingly. Otherwise, other developers might not be aware of the change, and they could end up writing code that doesn’t work as expected.\n",
    "\n",
    "    Finally, documenting exceptions can help with debugging. If we know what exceptions a function can raise, then we can narrow down the possible causes of an error when one occurs.  \n",
    "\n",
    "6. Provide Contextual Information When Raising an Exception :  \n",
    "    When an exception is raised, the Python interpreter stops execution of the program and prints out a traceback. The traceback starts with the line where the exception was raised and includes the lines of code that were executed leading up to that point.\n",
    "    If we provide contextual information when raising an exception, it will be included in the traceback and will help the person who is debugging the code to understand what went wrong.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "196fa007-3932-4eec-8480-2df19c2e7fba",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
